torch/cuda/comm.py

```diff
@@ -119,13 +131,20 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
         A tuple of tensors containing an elementwise sum of each group of
         inputs, placed on the ``destination`` device.
     """
+    dense_tensors = []  # shape (num_tensors, num_gpus)
     output = []
-    itrs = [_take_tensors(tensors, buffer_size) for tensors in inputs]
+    for tensor_at_gpus in zip(*inputs):
+        if tensor_at_gpus[0].is_sparse:
+            result = reduce_add(tensor_at_gpus, destination)
+            output.append(result)
+        else:
+            dense_tensors.append(tensor_at_gpus)
+    itrs = [_take_tensors(tensors, buffer_size) for tensors in zip(*dense_tensors)]
     for chunks in zip(*itrs):
-        flattened = [_flatten_tensors(chunk) for chunk in chunks]
-        result = reduce_add(flattened, destination)
-        output.extend(_unflatten_tensors(result, chunks[0]))
-    return tuple(output)
+        tensors = [_flatten_dense_tensors(chunk) for chunk in chunks]
```

Comment:
It would also be nice to avoid transposing Python data structures 3 times (interpreter overhead), but this seems more complicated.